master/buildbot/spec/types/build_data.raml

```diff
@@ -0,0 +1,47 @@
+#%RAML 1.0 DataType
+displayName: Build data
+description: |
+    This resource represents a key-value data pair associated to a build.
+    A build can have any number of key-value pairs.
+    The data is intended to be used for temporary purposes, until the build and all actions associated to it (such as reporters) are finished.
+
+    The value is a binary of potentially large size.
+    Certain APIs allow to retrieve the list of keys set to a build without fetching the values.
+
+    Update Methods
+    --------------
+
+    All update methods are available as attributes of ``master.data.updates``.
+
+    .. py:class:: buildbot.data.build_data.BuildData
+
+        .. py:method:: setBuildData(buildid, name, value, source)
+
+            :param integer buildid: build id to attach data to
+            :param unicode name: the name of the data
+            :param bytestr value: the value of the data as ``bytes``.
```

Comment:
how does that transform in json? I think json strings are utf-8, so this will clash if we encode binary data. I see that we also have ss patches that are binary, but I am not sure we have thorough test coverage on this.

Reply:
Why should they clash? If we store binary then we assume that the reader/writer knows the encoding and uses the same each time.

Reply:
Were you referring to use BLOB in the DB just as a workaround for some users using databases with non-UTF-8 default encoding for Text column?

Reply:
I was rather thinking of the case where a build_data generator would generator other data format than json (e.g zip containing several files). In this case, the byte string will contain arbitrary bytes.

I realize I wasn't clear in my first comment. I was there thinking of the REST api, which will convert data api's dicts into json. There is a toJson helper in buildbot.util, which converts datatimes into timestamps, but nothing to convert bytes. It will be needed.

```
json.dumps({'a': b'a'})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/3.7/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bytes is not JSON serializable
```


https://stackoverflow.com/questions/1443158/binary-data-in-json-string-something-better-than-base64